STYLEGUIDE.md

```diff
@@ -0,0 +1,716 @@
+interface IFooBar {
+  ...
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use pascal case without prefix */
+interface FooBar {
+  ...
+}
+```
+
+### Use pascal case for enums and enum properties
+Pascal case is common in most JavaScript/TypeScript projects for enums and enum properties.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use upper case snake case for enums and enum properties */
+enum SomeEnum {
+  FOO_BAR,
+  BAR_FOO
+}
+
+/* Don't use camel case for enums and enum properties */
+enum SomeEnum {
+  fooBar,
+  barFoo
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use pascal case */
+enum SomeEnum {
+  FooBar,
+  BarFoo
+}
+```
+### Use verbs in function names
+Every function should do something and that action should be described with a verb.
+The verb should fit the return type and value.
+An exception to this can be methods, that operate on an instance directly, e. g. `instance.toString()`, `instance.asObject()`.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use functions without verb */
+function toFoo(): Foo {
+  ...
+}
+
+/* Don't use functions with misleading verbs */
+function isFoo(): void {
+  ...
+  throw Error(...);
+}
+
+/* Don't use camel case for enums and enum properties */
+enum SomeEnum {
+  fooBar,
+  barFoo
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use functions with verbs */
+function transformToFoo(): Foo {
+  ...
+}
+
+/* Use functions with verbs that fit the return type */
+function isFoo(): boolean {
+  ...
+}
+
+/* Use functions with verbs that fit the return type */
+function validateFoo(): void {
+  ...
+  throw Error(...);
+}
+
+/* Exception, you can omit verbs in methods, that transform the instance that they are invoked on */
+class FooBar {
+  toString(): string {
+    ...
+  }
+
+  asObject(): Record<string, any> {
+    ...
+  }
+}
+```
+
+## Strings
+
+### Use single quotes ✓
+
+Use single quotes, unless not possible otherwise, e. g. when your string
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use double quotes */
+const foo = "foo";
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use single quotes */
+const foo = 'foo';
+
+/* Use double quotes if the quoted string contains single quotes. */
+const foo = "'foo'";
+```
+
+### String concatenation
+When possible use template literals, this makes the code more readable in most cases.
+Feel free to concatenate strings with `+` in case it makes to code better readable.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't concatenate strings with `+` */
+const foo = 'foo ' + bar + ' bar';
+
+/* Grey area: If template literals become too complex, don't use them. */
+const foo = `foo ${fn(`${bar}bar`)}`;
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Concatenate strings using template literals */
+const foo = `foo ${bar} bar`;
+
+/* Grey area: If template literals become too complex, feel free to concatenate with +. */
+const foo = `foo ${fn(bar +'bar')}`;
```

Comment:
```suggestion
const suffix = fn(bar + 'bar');
const foo = `foo ${suffix}`;
```

Reply:
This example was meant to produce a case where is a bit harder to read if there is a template string. Extracting it to a separate variable defeats the purpose a bit, because you could also write
```ts
fn(`${bar}bar`)
```

This was one of the rules where I had difficulties describing them, because the gray area is quite extensive and I wouldn't want to impose any hard rules.